library/alloc/tests/fmt.rs

```diff
@@ -1,7 +1,328 @@
+    format!("{}", a);
+    format!("{}", a);
+
+    // make sure that format! doesn't cause spurious unused-unsafe warnings when
+    // it's inside of an outer unsafe block
+    unsafe {
+        let a: isize = ::std::mem::transmute(3_usize);
+        format!("{}", a);
+    }
+
+    // test that trailing commas are acceptable
+    format!("{}", "test",);
+    format!("{foo}", foo = "test",);
+}
+
+// Basic test to make sure that we can invoke the `write!` macro with an
+// fmt::Write instance.
+#[test]
+fn test_write() {
+    let mut buf = String::new();
+    write!(&mut buf, "{}", 3);
+    {
+        let w = &mut buf;
+        write!(w, "{foo}", foo = 4);
+        write!(w, "{}", "hello");
+        writeln!(w, "{}", "line");
+        writeln!(w, "{foo}", foo = "bar");
+        w.write_char('☃');
+        w.write_str("str");
+    }
+
+    t!(buf, "34helloline\nbar\n☃str");
+}
+
+// Just make sure that the macros are defined, there's not really a lot that we
+// can do with them just yet (to test the output)
+#[test]
+fn test_print() {
+    print!("hi");
+    print!("{:?}", vec![0u8]);
+    println!("hello");
+    println!("this is a {}", "test");
+    println!("{foo}", foo = "bar");
+}
+
+// Just make sure that the macros are defined, there's not really a lot that we
+// can do with them just yet (to test the output)
+#[test]
+fn test_format_args() {
+    let mut buf = String::new();
+    {
+        let w = &mut buf;
+        write!(w, "{}", format_args!("{}", 1));
+        write!(w, "{}", format_args!("test"));
+        write!(w, "{}", format_args!("{test}", test = 3));
+    }
+    let s = buf;
+    t!(s, "1test3");
+
+    let s = fmt::format(format_args!("hello {}", "world"));
+    t!(s, "hello world");
+    let s = format!("{}: {}", "args were", format_args!("hello {}", "world"));
+    t!(s, "args were: hello world");
+}
+
+#[test]
+fn test_order() {
+    // Make sure format!() arguments are always evaluated in a left-to-right
+    // ordering
+    fn foo() -> isize {
+        static mut FOO: isize = 0;
+        unsafe {
+            FOO += 1;
+            FOO
+        }
+    }
+    assert_eq!(
+        format!("{} {} {a} {b} {} {c}", foo(), foo(), foo(), a = foo(), b = foo(), c = foo()),
+        "1 2 4 5 3 6".to_string()
+    );
+}
+
+#[test]
+fn test_once() {
+    // Make sure each argument are evaluated only once even though it may be
+    // formatted multiple times
+    fn foo() -> isize {
+        static mut FOO: isize = 0;
+        unsafe {
+            FOO += 1;
+            FOO
+        }
+    }
+    assert_eq!(format!("{0} {0} {0} {a} {a} {a}", foo(), a = foo()), "1 1 1 2 2 2".to_string());
+}
+
+#[test]
+fn test_refcell() {
+    let refcell = RefCell::new(5);
+    assert_eq!(format!("{:?}", refcell), "RefCell { value: 5 }");
+    let borrow = refcell.borrow_mut();
+    assert_eq!(format!("{:?}", refcell), "RefCell { value: <borrowed> }");
+    drop(borrow);
+    assert_eq!(format!("{:?}", refcell), "RefCell { value: 5 }");
+}
```

Comment:
Let's update this note as well:

https://github.com/rust-lang/rust/blob/3a4da87f58099f08620a9a3e812abd77301cafb2/library/core/src/fmt/mod.rs#L2241-L2242